rfcs/1-view-change-optimizations.md

```diff
@@ -0,0 +1,136 @@
+
+One more issue to consider with the suggestion from the paper is that while transferring state, we don't want the other replicas to time out and start a new view change. This could be a problem if the state to transfer is large and it takes longer than the view_change timeout to transfer all the data. A fix to this problem is to have the proposed primary periodically send `ContinueWaitForStartView` messages while it is updating it's state, but before it becomes the primary. This introduces a new message to the protocol, but should be safe as it is just a keep alive. This particular solution also has a problem, in that now the system is delayed for a very long view change. Maybe we should allow another view change to take place while the old proposed primary is still catching up, but this raises another question. What happens when we introduce streaming `NewState` messages for large state transfers. Do these operate in line with the consensus protocol? Once the stream has begun, does that replica go back to participating normally? Can it elect a new leader while an on-going `NewState` transfer is occurring?
+
+I favor the solution in this RFC to the solution in the paper, because it doesn't require extra worrying about timeouts during view change and doesn't require state transfer from a replica in view change to the primary. All `GetState` requests are answered from replicas in normal mode. Furthermore, the solution in the paper doesn't even mention preventing shipping the entire log in `StartView` messages. In the original protocol this log is shipped to all other nodes, whereas in this RFC, the minimal state is transferred to `F` nodes, and only remaining nodes *might* participate in state transfer. We could probably use the solution for that problem from this RFC though, even if we kept the rest of the changes in the paper.
+
+A second alternative not mentioned in the paper is to have the primary send the `epoch`, `view`, `op`, `commit_num`, and `last_normal_view` that it knows of for all replicas in every single prepare message, or every nth message. Then view change could proceed as normal, but each replica would only send the maximum possible missing messages in `DoViewChange` to the proposed primary from what they learned earlier. The proposed primary could then send only the maximum amount of possible missing messages to the replicas in the `StartView` message. This has the overhead of adding 40 bytes to every prepare during normal processing, which could add up to a lot of extra network usage. It also requires a bit more state to send during view change, since all nodes are now sending `log_tail`s to the proposed primary in `DoViewChange` messages. It also has the pathological case where a long partitioned node requires significant state transfer when becoming the primary. It does however have the benefit of mainly sticking with the protocol in the paper and avoiding unknowns with the other changes. It is also quite simple.
```

Comment:
Thanks for taking a look @evanmcc. The backpressure thing is a good idea. I also like this version for it's simplicity, and am sure we can reduce the cost. I know you mentioned something similar offhand to me the other day, but it didn't really hit me until I was more than halfway done with the RFC. It may very well turn out to be the right choice, in which case it should probably be it's own RFC.